### PR TITLE
[istio] get rid of revisions

### PR DESCRIPTION
**Pull request by fl64**
## Description
- removed istio 1.10.1 and related hacks
- refactored versions and revisions logic
- metric `d8_istio_pod_revision` renamed to `d8_istio_dataplane_metadata`
- metric `d8_istio_dataplane_metadata` enriched with `version`, `full_version`, `desired_version` and `desired_full_version` labels
- alerts were refactored by using the new handy metric

## Why do we need it, and what problem does it solve?
Labels with versions of sidecars will be needed to implement the feature https://github.com/deckhouse/deckhouse/issues/2144
Moreover, there were many controversies in docs and in the code. They aren't now.

## What is the expected result?
`d8_istio_pod_dataplane_metadata` with `version`, `full_version`, `desired_version` and `desired_full_version` labels

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: istio
type: chore
summary: Got rid of 1.10.1 version. Metric `d8_istio_pod_revision` renamed to `d8_istio_dataplane_metadata` and enriched with `version`, `full_version`, `desired_version` and `desired_full_version` labels. Alerts were refactored using the new metric. Also, a great refactoring for using term `version` instead of `revision`.
impact_level: default
```

### CODE REVIEW

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 1]
```diff
@@ -20,15 +20,14 @@ import (
...
 
 const (
-	istioRevsionAbsent = "absent"
-)
-
-var (
-	revisionsMonitoringMetricsGroup = "revisions"
+	istioRevsionAbsent           = "absent"
+	istioVersionAbsent           = "absent"
+	istioPodMetadataMetricName   = "d8_istio_pod_dataplane_metadata"
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:25]: d8_istio_pod_dataplane_metadata -> d8_istio_dataplane_metadata

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 2]
```diff
@@ -92,7 +91,8 @@ type IstioPodStatus struct {
 type IstioPodInfo struct {
 	Name             string
 	Namespace        string
-	Revision         string
+	Version          string // currently used
+	Revision         string // currently used
 	SpecificRevision string
 	InjectAnnotation bool
 	InjectLabel      bool
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:98]:
```suggestion
	SpecificRevision string // istio.io/rev: v1x13 label if it is
	InjectAnnotation bool   // sidecar.istio.io/inject: true annotation if it is
	InjectLabel      bool   // sidecar.istio.io/inject: true label if it is
```

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 3]
```diff
@@ -145,6 +145,16 @@ func (p *IstioDrivenPod) getIstioSpecificRevision() string {
 	return ""
 }
 
+func (p *IstioDrivenPod) getIstioVersion() string {
+	if specificPodVersion, ok := p.Annotations["istio.deckhouse.io/injected-sidecar-version"]; ok {
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:149]:
```suggestion
	if specificPodVersion, ok := p.Annotations["istio.deckhouse.io/version"]; ok {
```

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 4]
```diff
@@ -164,12 +175,19 @@ func applyIstioPodFilter(obj *unstructured.Unstructured) (go_hook.FilterResult,
...
 
-func revisionsMonitoring(input *go_hook.HookInput) error {
+func dataplaneMetadataExporter(input *go_hook.HookInput) error {
 	if !input.Values.Get("istio.internal.globalRevision").Exists() {
 		return nil
 	}
 
-	input.MetricsCollector.Expire(revisionsMonitoringMetricsGroup)
+	revisionFullVersionMap := make(map[string]string, 0)
+	if input.Values.Get("istio.internal.revisionFullVersionMap").Exists() {
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:184]: Let's trust in OpenAPI.

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 5]
```diff
@@ -211,13 +229,20 @@ func revisionsMonitoring(input *go_hook.HookInput) error {
 			continue
 		}
 
+		desiredVersion, ok := revisionFullVersionMap[desiredRevision]
+		if !ok {
+			desiredVersion = istioVersionAbsent
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:234]:
```suggestion
			desiredVersion = istioVersionUnknown
```

File: ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml [hunk 6]
```diff
@@ -9,14 +9,14 @@
...
-      plk_create_group_if_not_exists__d8_istio_revisions_misconfigurations: D8IstioRevisionsMisconfigurations,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
-      plk_grouped_by__d8_istio_revisions_misconfigurations: D8IstioRevisionsMisconfigurations,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
+      plk_create_group_if_not_exists__d8_istio_dataplane_misconfigurations: D8IstioDataplaneMisconfigurations,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
+      plk_grouped_by__d8_istio_dataplane_misconfigurations: D8IstioDataplaneMisconfigurations,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
       plk_markup_format: markdown
       plk_protocol_version: "1"
       summary: Desired control-plane version isn't installed
     expr: |
       (
-        d8_istio_pod_revision{revision!="absent"}
```
Comment (apolovov) [inline ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml:19]: There should be only labels that we need for the alert.

File: ee/modules/110-istio/openapi/values.yaml [hunk 7]
```diff
@@ -6,6 +6,16 @@ properties:
...
+      revisionFullVersionMap:
+        type: object
+        additionalProperties:
+          type: string
+        default:
+          "v1x13": "1.13.7"
+          "v1x12": "1.12.6"
+          "v1x10x1": "1.10.1"
+        x-examples:
+        - {"1x77": "1.77.7"}
```
Comment (apolovov) [inline ee/modules/110-istio/openapi/values.yaml:18]:
```suggestion
        - {"v1x77": "1.77.7"}
```

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go (hunk 1 above)
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter.go]: fixed

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go (hunk 3 above)
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:149]: fixed

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go (hunk 4 above)
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:184]: fixed

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go (hunk 5 above)
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:234]: fixed

File: ee/modules/110-istio/openapi/values.yaml (hunk 7 above)
Comment (fl64) [reply ee/modules/110-istio/openapi/values.yaml:18]: fixed

File: ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml (hunk 6 above)
Comment (fl64) [reply ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml:19]: fixed

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 8]
```diff
@@ -92,10 +92,11 @@ type IstioPodStatus struct {
 type IstioPodInfo struct {
 	Name             string
 	Namespace        string
-	Revision         string
-	SpecificRevision string
-	InjectAnnotation bool
-	InjectLabel      bool
+	Version          string // istio dataplane version
+	Revision         string // istio dataplane revision
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:96]:
```suggestion
	Version          string // istio dataplane version (i.e. "1.15.6")
	Revision         string // istio dataplane revision (i.e. "v1x15")
```

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 9]
```diff
@@ -92,10 +92,11 @@ type IstioPodStatus struct {
...
+	SpecificRevision string // istio.io/rev: vXxYZ label if it is
+	InjectAnnotation bool   // sidecar.istio.io/inject: true annotation if it is
+	InjectLabel      bool   // sidecar.istio.io/inject: true label if it is
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:99]:
```suggestion
	InjectAnnotation bool   // sidecar.istio.io/inject annotation if it is
	InjectLabel      bool   // sidecar.istio.io/inject label if it is
```

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 10]
```diff
@@ -145,6 +146,16 @@ func (p *IstioDrivenPod) getIstioSpecificRevision() string {
 	return ""
 }
 
+func (p *IstioDrivenPod) getIstioVersion() string {
+	if specificPodVersion, ok := p.Annotations["istio.deckhouse.io/version"]; ok {
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:150]:
```suggestion
	if istioVersion, ok := p.Annotations["istio.deckhouse.io/version"]; ok {
```

**Comment by apolovov** [comment]
1. image pilot-v1x13 -> pilot-v1x13x7 (and proxyv2 and operator)
2. OnStartup hook to convert "pilot-v1x13x7" -> internal: `{"versionMap":{"1.13": {"fullVersion": "1.13.7", "revision":"v1x13", "imageSuffix": "v1x13x7"}}}`
3. refactor `revisionsToInstall` to `versionsToInstall`, `operatorRevisionsToInstall` to `operatorVersionsToInstall`, `globalRevision` to `globalVersion`.

File: ee/modules/110-istio/hooks/discovery_operator_versions_to_install.go [hunk 11]
```diff
@@ -51,39 +52,34 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
-		if !internal.Contains(supportedRevisions, rev) {
-			unsupportedRevisions = append(unsupportedRevisions, rev)
+		if !internal.Contains(operatorVersionsToInstall, iopVer) {
+			operatorVersionsToInstall = append(operatorVersionsToInstall, iopVer)
 		}
 	}
+
 	if len(unsupportedRevisions) > 0 {
 		sort.Strings(unsupportedRevisions)
 		return fmt.Errorf("unsupported revisions: [%s]", strings.Join(unsupportedRevisions, ","))
```
Comment (fl64) [inline ee/modules/110-istio/hooks/discovery_operator_versions_to_install.go:78]: We need to discuss

File: ee/modules/110-istio/hooks/ensure_crds_istio.go [hunk 12]
```diff
@@ -17,25 +15,21 @@ import (
...
+	globalVersion := input.Values.Get("istio.internal.globalVersion").String()
 	var additionalVersions = make([]string, 0)
 	for _, versionResult := range input.ConfigValues.Get("istio.additionalVersions").Array() {
 		additionalVersions = append(additionalVersions, versionResult.String())
 	}
 
-	for _, versionResult := range input.Values.Get("istio.internal.supportedVersions").Array() {
-		version := versionResult.String()
+	for versionResult := range input.Values.Get("istio.internal.versionMap").Map() {
+		version := versionResult
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/ensure_crds_istio.go:32]:
```suggestion
	for version := range input.Values.Get("istio.internal.versionMap").Map() {
```

File: ee/modules/110-istio/hooks/ensure_crds_istio.go [hunk 13]
```diff
@@ -17,25 +15,21 @@ import (
...
-	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10}, // Order matters — we need globalVersion from discovery_revisions.go
+	OnStartup:    &go_hook.OrderedConfig{Order: 10}, // Order matters — we need globalVersion from discovery_versions_to_install.go
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10}, // Order matters — we need globalVersion from discovery_versions_to_install.go
 }, dependency.WithExternalDependencies(ensureCRDs))
 
 func ensureCRDs(input *go_hook.HookInput, dc dependency.Container) error {
 	var theNewestVersion string
 
-	var globalVersion string
-	if !input.Values.Exists("istio.internal.globalVersion") {
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/ensure_crds_istio.go:28]: Return the check pls.

File: ee/modules/110-istio/hooks/ensure_crds_istio.go [hunk 14]
```diff
@@ -17,25 +15,21 @@ import (
...
 	for _, versionResult := range input.ConfigValues.Get("istio.additionalVersions").Array() {
 		additionalVersions = append(additionalVersions, versionResult.String())
 	}
 
-	for _, versionResult := range input.Values.Get("istio.internal.supportedVersions").Array() {
-		version := versionResult.String()
+	for versionResult := range input.Values.Get("istio.internal.versionMap").Map() {
+		version := versionResult
 		if version == globalVersion || internal.Contains(additionalVersions, version) {
 			theNewestVersion = version
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/ensure_crds_istio.go:34]: if version > theNewestVersion — set theNewestVersion = version

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter_test.go [hunk 15]
```diff
@@ -149,6 +173,23 @@ var _ = Describe("Istio hooks :: revisions_monitoring ::", func() {
...
+		Entry("NS with definite revision, but revision is absent in revisionFullVersionMap",
+			[]string{
+				istioNsYAML(nsParams{
+					DefiniteRevision: "v1x00",
+				}),
+				istioPodYAML(podParams{
+					InjectionLabel:      true,
+					InjectionLabelValue: true,
+					CurrentRevision:     "v1x00",
+					Version:             "",
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/dataplane_metadata_exporter_test.go:185]:
```suggestion
					Version:             "", // annotation is absent
```

File: ee/modules/110-istio/hooks/discovery_versions_test.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+		})
+
+		It("Hook must execute successfully", func() {
+			Expect(f).To(ExecuteSuccessfully())
+			Expect(f.LogrusOutput.Contents()).To(HaveLen(0))
+			Expect(f.ValuesGet("istio.internal.versionMap").Map()).To(HaveLen(0))
+		})
+	})
+
+	Context("Some istio images exists", func() {
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/discovery_versions_test.go:31]:
```suggestion
	Context("Some istio images exist", func() {
```

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/hooks/ensure_crds_istio.go (hunk 14 above)
Comment (fl64) [reply ee/modules/110-istio/hooks/ensure_crds_istio.go]: fixed

File: ee/modules/110-istio/hooks/ensure_crds_istio.go [hunk 17]
```diff
@@ -7,39 +7,49 @@ package hooks
...
 var _ = sdk.RegisterFunc(&go_hook.HookConfig{
-	OnStartup:    &go_hook.OrderedConfig{Order: 10}, // Order matters — we need globalVersion from discovery_revisions.go
-	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10}, // Order matters — we need globalVersion from discovery_revisions.go
+	OnStartup:    &go_hook.OrderedConfig{Order: 10}, // Order matters — we need globalVersion from discovery_versions_to_install.go
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10}, // Order matters — we need globalVersion from discovery_versions_to_install.go
 }, dependency.WithExternalDependencies(ensureCRDs))
 
 func ensureCRDs(input *go_hook.HookInput, dc dependency.Container) error {
-	var theNewestVersion string
+	// collect all istio versions global + additional
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/ensure_crds_istio.go:26]: fixed

File: ee/modules/110-istio/hooks/ensure_crds_istio.go [hunk 18]
```diff
@@ -7,39 +7,49 @@ package hooks
...
+		if !internal.Contains(istioVersions, versionResult.String()) {
+			istioVersions = append(istioVersions, versionResult.String())
+		}
 	}
 
-	for _, versionResult := range input.Values.Get("istio.internal.supportedVersions").Array() {
-		version := versionResult.String()
-		if version == globalVersion || internal.Contains(additionalVersions, version) {
-			theNewestVersion = version
+	// semvers is a slice for sorting semver
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/ensure_crds_istio.go:41]: fixed

File: ee/modules/110-istio/hooks/internal/istio_versions/istio_versions.go [hunk 19]
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2022 Flant JSC
+Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+*/
+
+package istio_versions
+
+import "encoding/json"
+
+type IstioVersionsMapType map[string]IstioVersionInfo
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/internal/istio_versions/istio_versions.go:10]: fixed

File: ee/modules/110-istio/templates/control-plane/iop.yaml [hunk 20]
```diff
@@ -123,7 +127,7 @@ spec:
       - d8-istio-sidecar-registry
 
       proxy:
-        image: {{ include "helm_lib_module_image" (list $ (printf "proxyv2%s" ($revision | title))) }}
+        image: {{ include "helm_lib_module_image" (list $ (printf "proxyv2%s" ($imageSuffix | title))) }}
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/templates/control-plane/iop.yaml:130]: fixed

File: ee/modules/110-istio/templates/control-plane/service-global.yaml [hunk 21]
```diff
@@ -1,13 +1,16 @@
...
 # Workaround for https://github.com/istio/istio/issues/27501
 apiVersion: v1
 kind: Service
 metadata:
   name: istiod
   namespace: d8-istio
   annotations:
-    istio.deckhouse.io/global-version: {{ .Values.istio.internal.globalVersion | quote }}
-  {{- include "helm_lib_module_labels" (list . (dict "app" "istiod" "istio" "pilot" "istio.io/rev" .Values.istio.internal.globalRevision)) | nindent 2 }}
+    istio.deckhouse.io/global-version: {{ $revision | quote }}
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/templates/control-plane/service-global.yaml:12]: fixed

File: ee/modules/110-istio/hooks/discovery_versions_to_install.go [hunk 22]
```diff
@@ -72,35 +66,34 @@ func revisionsDiscovery(input *go_hook.HookInput, dc dependency.Container) error
...
+		}
 	}
 
 	if len(unsupportedVersions) > 0 {
 		sort.Strings(unsupportedVersions)
 		return fmt.Errorf("unsupported versions: [%s]", strings.Join(unsupportedVersions, ","))
 	}
 
-	sort.Strings(revisionsToInstall)
+	sort.Strings(versionsToInstall)
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/discovery_versions_to_install.go:93]: added

File: ee/modules/110-istio/hooks/discovery_versions_test.go (hunk 16 above)
Comment (fl64) [reply ee/modules/110-istio/hooks/discovery_versions_test.go:31]: fixed

File: ee/modules/110-istio/hooks/discovery_operator_versions_to_install.go [hunk 23]
```diff
@@ -51,39 +52,34 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
-		supportedRevisions = append(supportedRevisions, internal.VersionToRevision(versionResult.String()))
-	}
+	versionMap := istio_versions.VersionMapStrToVersionMapType(input.Values.Get("istio.internal.versionMap").String())
 
-	var revisionsToInstallResult = input.Values.Get("istio.internal.revisionsToInstall").Array()
-	for _, revisionResult := range revisionsToInstallResult {
-		operatorRevisionsToInstall = append(operatorRevisionsToInstall, revisionResult.String())
+	var versionsToInstallResult = input.Values.Get("istio.internal.versionsToInstall").Array()
+	for _, revisionResult := range versionsToInstallResult {
+		operatorVersionsToInstall = append(operatorVersionsToInstall, revisionResult.String())
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/discovery_operator_versions_to_install.go:62]: fixed

File: ee/modules/110-istio/hooks/discovery_operator_versions_to_install.go [hunk 24]
```diff
@@ -51,39 +52,34 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
 	}
 
 	for _, iop := range input.Snapshots["istiooperators"] {
 		iopInfo := iop.(IstioOperatorCrdInfo)
-		if !internal.Contains(operatorRevisionsToInstall, iopInfo.Revision) {
-			operatorRevisionsToInstall = append(operatorRevisionsToInstall, iopInfo.Revision)
+		iopVer := versionMap.GetVersionByRevision(iopInfo.Revision)
+		if iopVer == "" {
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/discovery_operator_versions_to_install.go:68]: done

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 25]
```diff
@@ -145,6 +147,16 @@ func (p *IstioDrivenPod) getIstioSpecificRevision() string {
 	return ""
 }
 
+func (p *IstioDrivenPod) getIstioVersion() string {
+	if istioVersion, ok := p.Annotations["istio.deckhouse.io/version"]; ok {
+		if istioVersion == "" {
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:152]: ok!

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 26]
```diff
@@ -145,6 +147,16 @@ func (p *IstioDrivenPod) getIstioSpecificRevision() string {
 	return ""
 }
 
+func (p *IstioDrivenPod) getIstioVersion() string {
+	if istioVersion, ok := p.Annotations["istio.deckhouse.io/version"]; ok {
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:151]: done

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter.go [hunk 27]
```diff
@@ -211,13 +225,20 @@ func revisionsMonitoring(input *go_hook.HookInput) error {
...
+		if desiredFullVersion == "" {
+			desiredFullVersion = istioVersionUnknown
+		}
+
 		labels := map[string]string{
 			"namespace":        istioPodInfo.Namespace,
 			"dataplane_pod":    istioPodInfo.Name,
 			"desired_revision": desiredRevision,
 			"revision":         istioPodInfo.Revision,
+			"version":          istioPodInfo.Version,
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter.go:238]: fixed

File: ee/modules/110-istio/hooks/dataplane_metadata_exporter_test.go [hunk 28]
```diff
@@ -149,6 +173,23 @@ var _ = Describe("Istio hooks :: revisions_monitoring ::", func() {
 					InjectionLabelValue: false,
 				}),
 			}, nil),
+		Entry("NS with definite revision, but revision is absent in revisionFullVersionMap",
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/hooks/dataplane_metadata_exporter_test.go:176]: there is already test case "NS global revision, pod revision is absent (no sidecar)"

File: ee/modules/110-istio/templates/control-plane/service-global.yaml [hunk 29]
```diff
@@ -1,13 +1,16 @@
...
+  {{- include "helm_lib_module_labels" (list . (dict "app" "istiod" "istio" "pilot" "istio.io/rev" $revision)) | nindent 2 }}
```
[the remark this answers is not in the document]
Comment (fl64) [reply ee/modules/110-istio/templates/control-plane/service-global.yaml:13]: replaced

File: ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml [hunk 30]
```diff
@@ -0,0 +1,176 @@
+- name: d8.istio.dataplane
+  rules:
+  - alert: D8IstioActualVersionIsNotInstalled
+    annotations:
+      description: |
+        There are pods with injected sidecar with version `{{$labels.version}}` and revision `{{$labels.revision}}` in namespace `{{$labels.namespace}}`, but the control-plane version isn't installed. Consider installing it or change the Namespace or Pod configuration.
```
Comment (apolovov) [inline ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml:6]:
```suggestion
        There are pods with injected sidecar with version `{{$labels.version}}` (revision `{{$labels.revision}}`) in namespace `{{$labels.namespace}}`, but the control-plane version isn't installed. Consider installing it or change the Namespace or Pod configuration.
```

File: ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml [hunk 31]
```diff
@@ -0,0 +1,176 @@
...
+        )
+      )
+    for: 5m
+    labels:
+      severity_level: "4"
+      tier: cluster
+  - alert: D8IstioDesiredVersionIsNotInstalled
+    annotations:
+      description: |
+        There is desired istio control plane version `{{$labels.desired_version}}` and revision `{{$labels.revision}}` configured for pods in namespace `{{$labels.namespace}}`, but the version isn't installed. Consider installing it or change the Namespace or Pod configuration.
```
Comment (apolovov) [inline ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml:37]:
```suggestion
        There is desired istio control plane version `{{$labels.desired_version}}` (revision `{{$labels.revision}}`) configured for pods in namespace `{{$labels.namespace}}`, but the version isn't installed. Consider installing it or change the Namespace or Pod configuration.
```

File: ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml [hunk 32]
```diff
@@ -0,0 +1,176 @@
...
+           d8_istio_dataplane_metadata{desired_revision="absent",revision!="absent"}
+        )
+    for: 5m
+    labels:
+      severity_level: "4"
+      tier: cluster
+  - alert: D8IstioPodsWithoutIstioSidecar
+    annotations:
+      description: |
+        There are pods `{{$labels.dataplane_pod}}` in `{{$labels.namespace}}` Namespace without istio sidecars, but the istio-injection is configured.
```
Comment (apolovov) [inline ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml:97]:
```suggestion
        There is a Pod `{{$labels.dataplane_pod}}` in `{{$labels.namespace}}` Namespace without istio sidecars, but the istio-injection is configured.
```

File: ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml [hunk 33]
```diff
@@ -0,0 +1,176 @@
...
+        )
+    for: 5m
+    labels:
+      severity_level: "4"
+      tier: cluster
+  - alert: D8IstioActualDataPlaneVersionNeDesired
+    annotations:
+      description: |
+        There are Pods in Namespace `{{$labels.namespace}}` with istio data-plane version `{{$labels.version}}`, but the desired one is `{{$labels.desired_version}}`.
+        Impact — revision is to change after Pod restarting.
```
Comment (apolovov) [inline ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml:120]:
```suggestion
        Impact — istio version is to change after Pod restarting.
```

File: ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml [hunk 34]
```diff
@@ -0,0 +1,176 @@
...
+        Impact — Pods have lost their sync with k8s state.
+        Getting orphaned pods:
+        ```
+        kubectl -n {{ $labels.namespace }} get pods -l 'service.istio.io/canonical-name' -o json | jq --arg revision {{ $labels.revision }} '.items[] | select(.metadata.annotations."sidecar.istio.io/status" // "{}" | fromjson | .revision == $revision) | .metadata.name'
+        ```
+      plk_create_group_if_not_exists__d8_istio_dataplane_misconfigurations: D8IstioDataplaneMisconfigurations,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
+      plk_grouped_by__d8_istio_dataplane_misconfigurations: D8IstioDataplaneMisconfigurations,tier=~tier,prometheus=deckhouse,kubernetes=~kubernetes
+      plk_markup_format: markdown
+      plk_protocol_version: "1"
+      summary: Desired control-plane version isn't installed
```
Comment (apolovov) [inline ee/modules/110-istio/monitoring/prometheus-rules/dataplane.yaml:16]:
```suggestion
      summary: control-plane version for Pod with already injected sidecar isn't installed
```

**Review by apolovov — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @apolovov. I've started the workflow for label 'e2e/run/openstack'...

**Pull request merged**